// @mysten/sui.js/dist/cjs/keypairs/secp256r1/publickey.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var publickey_exports = {};
__export(publickey_exports, {
  Secp256r1PublicKey: () => Secp256r1PublicKey
});
module.exports = __toCommonJS(publickey_exports);
var import_bcs = require("@mysten/bcs");
var import_blake2b = require("@noble/hashes/blake2b");
var import_utils = require("@noble/hashes/utils");
var import_types = require("../../types/index.js");
var import_publickey = require("../../cryptography/publickey.js");
var import_signature = require("../../cryptography/signature.js");
const SECP256R1_PUBLIC_KEY_SIZE = 33;
class Secp256r1PublicKey {
  /**
   * Create a new Secp256r1PublicKey object
   * @param value secp256r1 public key as buffer or base-64 encoded string
   */
  constructor(value) {
    if (typeof value === "string") {
      this.data = (0, import_bcs.fromB64)(value);
    } else if (value instanceof Uint8Array) {
      this.data = value;
    } else {
      this.data = Uint8Array.from(value);
    }
    if (this.data.length !== SECP256R1_PUBLIC_KEY_SIZE) {
      throw new Error(
        `Invalid public key input. Expected ${SECP256R1_PUBLIC_KEY_SIZE} bytes, got ${this.data.length}`
      );
    }
  }
  /**
   * Checks if two Secp256r1 public keys are equal
   */
  equals(publicKey) {
    return (0, import_publickey.bytesEqual)(this.toBytes(), publicKey.toBytes());
  }
  /**
   * Return the base-64 representation of the Secp256r1 public key
   */
  toBase64() {
    return (0, import_bcs.toB64)(this.toBytes());
  }
  /**
   * Return the byte array representation of the Secp256r1 public key
   */
  toBytes() {
    return this.data;
  }
  /**
   * Return the base-64 representation of the Secp256r1 public key
   */
  toString() {
    return this.toBase64();
  }
  /**
   * Return the Sui address associated with this Secp256r1 public key
   */
  toSuiAddress() {
    let tmp = new Uint8Array(SECP256R1_PUBLIC_KEY_SIZE + 1);
    tmp.set([import_signature.SIGNATURE_SCHEME_TO_FLAG["Secp256r1"]]);
    tmp.set(this.toBytes(), 1);
    return (0, import_types.normalizeSuiAddress)(
      (0, import_utils.bytesToHex)((0, import_blake2b.blake2b)(tmp, { dkLen: 32 })).slice(0, import_types.SUI_ADDRESS_LENGTH * 2)
    );
  }
  /**
   * Return the Sui address associated with this Secp256r1 public key
   */
  flag() {
    return import_signature.SIGNATURE_SCHEME_TO_FLAG["Secp256r1"];
  }
}
Secp256r1PublicKey.SIZE = SECP256R1_PUBLIC_KEY_SIZE;
