// @mysten/sui.js/dist/cjs/types/faucet.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var faucet_exports = {};
__export(faucet_exports, {
  FaucetCoinInfo: () => FaucetCoinInfo,
  FaucetResponse: () => FaucetResponse
});
module.exports = __toCommonJS(faucet_exports);
var import_superstruct = require("superstruct");
var import_common = require("./common.js");
const FaucetCoinInfo = (0, import_superstruct.object)({
  amount: (0, import_superstruct.number)(),
  id: import_common.ObjectId,
  transferTxDigest: import_common.TransactionDigest
});
const FaucetResponse = (0, import_superstruct.object)({
  transferredGasObjects: (0, import_superstruct.array)(FaucetCoinInfo),
  error: (0, import_superstruct.nullable)((0, import_superstruct.string)())
});
