// @mysten/sui.js/dist/esm/types/checkpoints.js
import {
  array,
  number,
  object,
  string,
  tuple,
  boolean,
  optional,
  any,
  nullable
} from "superstruct";
import { TransactionDigest, TransactionEffectsDigest } from "./common.js";
const GasCostSummary = object({
  computationCost: string(),
  storageCost: string(),
  storageRebate: string(),
  nonRefundableStorageFee: string()
});
const CheckPointContentsDigest = string();
const CheckpointDigest = string();
const ECMHLiveObjectSetDigest = object({
  digest: array(number())
});
const CheckpointCommitment = any();
const ValidatorSignature = string();
const EndOfEpochData = object({
  nextEpochCommittee: array(tuple([string(), string()])),
  nextEpochProtocolVersion: string(),
  epochCommitments: array(CheckpointCommitment)
});
const ExecutionDigests = object({
  transaction: TransactionDigest,
  effects: TransactionEffectsDigest
});
const Checkpoint = object({
  epoch: string(),
  sequenceNumber: string(),
  digest: CheckpointDigest,
  networkTotalTransactions: string(),
  previousDigest: optional(CheckpointDigest),
  epochRollingGasCostSummary: GasCostSummary,
  timestampMs: string(),
  endOfEpochData: optional(EndOfEpochData),
  // TODO(jian): remove optional after 0.30.0 is released
  validatorSignature: optional(ValidatorSignature),
  transactions: array(TransactionDigest),
  checkpointCommitments: array(CheckpointCommitment)
});
const CheckpointPage = object({
  data: array(Checkpoint),
  nextCursor: nullable(string()),
  hasNextPage: boolean()
});
export {
  CheckPointContentsDigest,
  Checkpoint,
  CheckpointCommitment,
  CheckpointDigest,
  CheckpointPage,
  ECMHLiveObjectSetDigest,
  EndOfEpochData,
  ExecutionDigests,
  GasCostSummary,
  ValidatorSignature
};
